src/obelisk_client.cpp

```diff
@@ -579,8 +640,8 @@ void obelisk_client::attach_handlers()
         const auto ec = source.read_error_code();
         if (ec)
         {
-            handler->second(ec, 0, 0, {});
-            update_handlers_.erase(handler);
+            handler(ec, 0, 0, {});
```

Comment:
A caller-provided handler should not be invoked under a lock unless absolutely necessary. This sets up a deadlock condition in the case where the handler reenters a critical section under the same mutex. Critical sections should be kept small enough that all possible code paths under lock can be visually-verified.